README.rst
DiProPerm
----

**author**: `Iain Carmichael`_

Additional documentation, examples and code revisions are coming soon.
For questions, issues or feature requests please reach out to Iain:
iain@unc.edu.

Overview
========

This package implements Direction-Projection-Permutation for High Dimensional
Hypothesis Tests (DiPoPerm). For details see Wei et al, 2016 (`paper link`_, `arxiv link`_). DiProPerm "rigorously assesses whether a binary linear classifier is detecting statistically significant differences between two high-dimensional distributions."



Wei, S., Lee, C., Wichers, L., & Marron, J. S. (2016). Direction-projection-permutation for high-dimensional hypothesis tests. Journal of Computational and Graphical Statistics, 25(2), 549-569.

Installation
============

The diproperm package can be installed via pip or github. This package is currently only tested in python 3.6.

::

    pip install diproperm


::

    git clone https://github.com/idc9/diproperm.git
    python setup.py install

Example
=======

.. code:: python

    from sklearn.datasets import make_blobs
    import numpy as np
    import matplotlib.pyplot as plt
    # %matplotlib inline

    from diproperm.DiProPerm import DiProPerm

    # toy binary class dataset (two isotropic Gaussians)
    X, y = make_blobs(n_samples=100, n_features=2, centers=2, cluster_std=2)

    # DiProPerm with mean difference classifier, mean difference summary
    # statistic, and 1000 permutation samples.
    dpp = DiProPerm(B=1000, separation_stats=['md', 't', 'auc'], clf='md')
    dpp.fit(X, y)

    dpp.test_stats_['md']

.. code:: python

    {'Z': 11.704865481794599,
     'cutoff_val': 1.2678333596648679,
     'obs': 4.542253375623943,
     'pval': 0.0,
     'rejected': True}

.. code:: python

    plt.figure(figsize=[12, 5])
    # show histogram of separation statistics
    plt.subplot(1, 2, 1)
    dpp.plot_perm_sep_stats(stat='md')

    # the observed scores
    plt.subplot(1, 2, 2)
    dpp.plot_observed_scores()

.. image:: doc/figures/dpp_plots.png


For more example code see `these example notebooks`_.

Help and Support
================

Additional documentation, examples and code revisions are coming soon.
For questions, issues or feature requests please reach out to Iain:
iain@unc.edu.

Documentation
^^^^^^^^^^^^^

The source code is located on github: https://github.com/idc9/diproperm

Testing
^^^^^^^

Testing is done using `nose`.

Contributing
^^^^^^^^^^^^

We welcome contributions to make this a stronger package: data examples,
bug fixes, spelling errors, new features, etc.



.. _Iain Carmichael: https://idc9.github.io/
.. _paper link: https://www.tandfonline.com/doi/abs/10.1080/10618600.2015.1027773
.. _arxiv link: https://arxiv.org/pdf/1304.0796.pdf
.. _these example notebooks: https://github.com/idc9/diproperm/tree/master/doc


changelog.md
# 11/5/18 (version 0.0.3)

Now saves permutation scores. A number of changes to the API and backend.

- save permutation scores
- new visualization diagnostics

# 10/30/18 (version 0.0.2)
- changes attribute names (now test_stats_, perm_samples_)
- added custom test statistics

# 10/30/18 (version 0.0.1)
- first upload


diproperm/viz.py
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter

from diproperm.viz_utils import jitter


def plot_observed_scores(obs_scores, y):
    """
    Plots a histogram of the observed scores.

    Parameters
    ----------
    obs_scores: array-like, shape (n_samples, )
        The observed scores from projecting the data onto the classification
        normal vector.

    y: array-like, shape (n_samples, )
        The observed class labels.
    """
    classes = np.unique(y)
    mask0 = y == classes[0]
    counts = Counter(y)

    label0 = '{} (n={})'.format(classes[0], counts[classes[0]])
    sns.distplot(obs_scores[mask0], color='red', label=label0)
    jitter(obs_scores[mask0], color='red')

    label1 = '{} (n={})'.format(classes[1], counts[classes[1]])
    sns.distplot(obs_scores[~mask0], color='blue', label=label1)
    jitter(obs_scores[~mask0], color='blue')
    plt.legend()
    plt.xlabel('observed scores')


def plot_perm_scores(perm_scores, perm_y, obs_y):
    """

    Parameters
    ----------
    perm_scores: array-like, shape (B, n_samples)
        The permutation scores.

    perm_scores: array-like, shape (B, n_samples)
        The permutation class labels.

    y_obs: array-like, shape (n_samples, )
        The observed class labels.

    """
    classes = np.unique(obs_y)

    obs0_perm0 = (obs_y == classes[0]) & (perm_y == classes[0])
    obs0_perm1 = (obs_y == classes[0]) & (perm_y == classes[1])
    obs1_perm0 = (obs_y == classes[1]) & (perm_y == classes[0])
    obs1_perm1 = (obs_y == classes[1]) & (perm_y == classes[1])

    perm0_color = 'red'
    perm1_color = 'blue'

    obs0_marker = 'o'
    obs0_size = 100
    obs1_marker = 'x'
    obs1_size = 150

    obs0_perm0_dist_kws = {'color': perm0_color, 'hist_kws': {'alpha': .1}}
    obs0_perm0_jitter_kws = {'color': perm0_color, 's': obs0_size, 'marker': obs0_marker}

    obs0_perm1_dist_kws = {'color': perm1_color, 'hist_kws': {'alpha': .1}}
    obs0_perm1_jitter_kws = {'color': perm1_color, 's': obs0_size, 'marker': obs0_marker}

    obs1_perm0_dist_kws = {'color': perm0_color, 'hist_kws': {'alpha': .1}}
    obs1_perm0_jitter_kws = {'color': perm0_color, 's': obs1_size, 'marker': obs1_marker}

    obs1_perm1_dist_kws = {'color': perm1_color, 'hist_kws': {'alpha': .1}}
    obs1_perm1_jitter_kws = {'color': perm1_color, 's': obs1_size, 'marker': obs1_marker}

    label_obs0_perm0 = 'obs {}, perm {}'.format(classes[0], classes[0])
    sns.distplot(perm_scores[obs0_perm0], label=label_obs0_perm0,
                 **obs0_perm0_dist_kws)
    jitter(perm_scores[obs0_perm0], **obs0_perm0_jitter_kws)

    label_obs0_perm1 = 'obs {}, perm {}'.format(classes[0], classes[1])
    sns.distplot(perm_scores[obs0_perm1], label=label_obs0_perm1,
                 **obs0_perm1_dist_kws)
    jitter(perm_scores[obs0_perm1], **obs0_perm1_jitter_kws)

    label_obs1_perm0 = 'obs {}, perm {}'.format(classes[1], classes[0])
    sns.distplot(perm_scores[obs1_perm0], label=label_obs1_perm0,
                 **obs1_perm0_dist_kws)
    jitter(perm_scores[obs1_perm0], **obs1_perm0_jitter_kws)

    label_obs1_perm1 = 'obs {}, perm {}'.format(classes[1], classes[1])
    sns.distplot(perm_scores[obs1_perm1], label=label_obs1_perm1,
                 **obs1_perm1_dist_kws)
    jitter(perm_scores[obs1_perm1], **obs1_perm1_jitter_kws)

    plt.xlabel('permutation scores')

    plt.legend()


def plot_perm_sep_stats(obs, perm, stat, rejected, p, Z, cuttoff, alpha, B, bins=30):
    """
    Plots a histogram of the DiProPerm distribution.

    Parameters
    ----------
    stat: str
        Which summary statistic to show.

    bins: int
        Number of bins for histogram.

    stat: str
        Name of the separation statistic.

    rejected: bool

    p: float

    Z: float

    cuttoff: float

    alpha: float

    bins: int

    """

    plt.hist(perm,
             color='black',
             label='permutation stats (B = {})'.format(B),
             bins=bins)

    jitter(perm, color='green')

    if rejected:
        obs_lw = 3
        sig_lab = 'significant'

    else:
        obs_lw = 1
        sig_lab = 'not significant'

    obs_label = 'obs stat ({}, p = {:1.3f}, Z = {:1.3f})'.\
                format(sig_lab, p, Z)

    plt.axvline(obs,
                color='red', lw=obs_lw,
                label=obs_label)

    plt.axvline(cuttoff, color='grey',
                ls='dashed',
                label='significance cutoff (alpha = {:1.3f})'.
                      format(alpha))

    plt.xlabel('DiProPerm {}'.format(stat))
    plt.legend()


diproperm/classifiers.py
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors.nearest_centroid import NearestCentroid


def get_training_fun(clf, param_grid=None, metric='roc_auc', n_splits=5):
    """
    Parameters
    ----------
    clf:
        linear classifier object following sklearn's API.

    param_grid: dict, None
        Parameters to cross-validate over. If None, will not perform cross-validation.

    metric: str
        Metric to use to select hyper-parameters. See scoring argument to GridSearchCV()

    n_splits: int
        Number of folds for cross validation

    Output
    ------
    get_normal_vector: callable
        Takes X, y input, trains the linear classifier then returns
        the normal vector.


        Parameters
        ----------
        X: array-like, shape (n_samples, n_features)
            The X data.

        y: array-like, shape (n_samples, )
            Class labels.
    """

    def get_normal_vector(X, y):
        """
        Selects hyper-parameters using cross-validation. Then refits
        using the full data set and returns the classification scores.

        Parameters
        -----------
        X: array-like, shape (n_samples, n_features)
            The X data.

        y: array-like, shape (n_samples, )
            Class labels.

        Output
        -----

        w: array-like, shape (n_features, )
            Normalized classification vector.
        """

        if param_grid is None or len(param_grid) == 0:

            cv = GridSearchCV(estimator=clf,
                              param_grid=param_grid,
                              scoring=metric,
                              cv=StratifiedKFold(n_splits=n_splits),
                              refit=True).fit(X, y)

            clf_trained = cv.best_estimator_
        else:
            clf_trained = clf.fit(X, y)

        w = get_clf_normal_vector(clf_trained).reshape(-1)
        w /= np.linalg.norm(w)
        return w

    return get_normal_vector


def get_clf_normal_vector(clf):
    """
    Returns the normal vector from a linear classifier object.

    Parameters
    ----------
    clf:
        Linear classifier object.
    """
    if hasattr(clf, 'coef_'):
        return clf.coef_

    elif type(clf) == GaussianNB:
        return get_GNB_direction(clf)

    elif type(clf) == NearestCentroid:
        return get_NC_direction(clf)

    else:
        return None


def get_NC_direction(clf):
    """
    Returns the normal vector for NearestCentroid with binary classes
    """
    assert type(clf) == NearestCentroid
    assert clf.centroids_.shape[0] == 2
    return (clf.centroids_[0, :] - clf.centroids_[1, :]).reshape(-1)


def get_GNB_direction(clf):
    """
    Returns the normal vector for GaussianNB with binary classes
    """
    assert type(clf) == GaussianNB
    assert clf.theta_.shape[0] == 2
    w_md = (clf.theta_[0, :] - clf.theta_[1, :]).reshape(-1)
    p0 = clf.class_prior_[0]
    p1 = clf.class_prior_[1]
    sigma = p0 * clf.sigma_[0, :] + p1 * clf.sigma_[1, :]  # TODO: double check
    Sigma_inv = np.diag(_safe_invert(sigma))
    w_nb = Sigma_inv.dot(w_md)
    return w_nb.reshape(-1)


def get_md_normal_vector(X, y):
    """
    Computes the mean difference normal vector, w_md, where
    w_md = mean(X_positive) - mean(X_negative)
    w_md = normalize(w_md)

    Parameters
    -----------
    X: array-like, shape (n_samples, n_features)
        The X data.

    y: array-like, shape (n_samples, )
        Class labels.

    Output
    -----

    w: array-like, shape (n_features, )
        Normalized classification vector.

    """
    y = np.array(y)
    X = np.array(X)
    if X.ndim == 1:
        X = X.reshape(-1, 1)
    classes = np.unique(y)
    assert len(classes) == 2

    w = X[y == classes[0], :].mean(axis=0) - X[y == classes[1], :].mean(axis=0)
    w /= np.linalg.norm(w)
    return w
    # return np.dot(X, w)


def _safe_invert(x):
    if np.allclose(x, 0):
        return 0
    else:
        return 1.0/x


diproperm/DiProPerm.py
import numpy as np
from scipy.stats import ttest_ind
from collections import Counter
import matplotlib.pyplot as plt
from sklearn.externals.joblib import dump, load
from sklearn.metrics import roc_auc_score
from sklearn.externals.joblib import Parallel, delayed

from diproperm.classifiers import get_md_normal_vector
from diproperm.viz import plot_perm_sep_stats, plot_observed_scores, \
    plot_perm_scores


class DiProPerm(object):
    """
    Direction-Projection-Permutation for High Dimensional Hypothesis Tests
    (DiProPerm). See (Wei et al, 2016) for details.

    Let X1,...,Xm and Y1,...,Yn be independent random samples of d dimensional
    random vectors with multivariate distributions F1 and F2, respectively.
    We are interested interested the null hypothesis of equal distributions
    H0 : F1 = F2 versus H1 : F1 != F2.

    DiProPerm does the following procedure once for the observed data (X, y)
    and then B times for permuted data (X, y_perm).

    - fit a linear classifier normal vector, w, (e.g. using the
    mean difference or SVM).

    - project the data, X, onto the normal vector to
    obtain the scores, Xw.

    - compute a separation statistic measuring how well separated the
    two class scores are (e.g. ROC-AUC, difference of class means).

    The B permuted separation statistics are samples from the null
    distribution and can be used to conduct a hypothesis test.


    Wei, S., Lee, C., Wichers, L., & Marron, J. S. (2016).
    Direction-projection-permutation for high-dimensional hypothesis tests.
    Journal of Computational and Graphical Statistics, 25(2), 549-569.

    Parameters
    ----------
    B: int
        Number of permutations to sample.

    clf: str, callable
        Linear classification algorithm used to find a direction.
        If str, must be one of ['md']. If callable, must take two arguments:
        X, y and return a vector of scores where the scores are Xw, w the
        linear classification normal vector.

    separation_stats: str, list {'md', 't', 'auc'}
        The univariate two-sample statistics measuring the separation
        between the two classes' scores. See section 2.2 of (Wei et al, 2016).
        Multiple can be provided as a list.


    custom_sep_stats: None, iterable of (str, callable) tuples
        Custom separation statistics which are a function of (scores, y).
        The str argument identifies the name of the statistic (used for
        keys of obs_sep_stats_, perm_sep_stats_ attributes).
        The function, f, must take arguments f(scores, y) and return a value.
        The larger the value, the more separated the two classes are.

    custom_tests_stats: None, iterable of (str, callable) tuples
        Custom test statistics which are a function of (obs_stat, perm_samples).
        The str argument identifies the name of the statistic (used for
        keys of test_stats_ attribute). The function, f, must take
        arguments f(obs_stat, perm_samples) and return a value.

    alpha: float
        Cutoff for significance.

    n_jobs: None, int
        Number of jobs for parallel processing permutations using
        sklearn.externals.joblib.Parallel. If None, will not use
        parallel processing.

    Attributes
    ----------

    test_stats_: dict
        dict keyed by summary statistics containing the test statistics
        (e.g. p-value, Z statistic, etc)

    obs_scores_: np.arrays, shape (n_samples, )
        The observed scores (i.e. projection of the training data onto
        the classification normal vector).

    perm_scores_: np.arrays, shape (B, n_samples)
        The permutation scores.

    perm_y: np.array, shape (B, n_samples)
        The permuted labels used for each permutation sample.

    obs_sep_stats_: dict of np.arrays, shape (n_samples, )
        dict containing the observed separation statistics. Keys are the
        separation statistics (e.g. 'md', 't', 'auc').

    perm_sep_stats_: dict of np.array, shape (B, n_samples)
        dict containing the permuted separation statistics.

    metadata_: dict

    obs_y_: array-like
        The observed class labels i.e. the y argument to fit

    classes_: list
        Class labels. For classification, classes_[0] is considered to
        be the positive class.

    """
    def __init__(self, B=100, clf='md',
                 separation_stats=['md', 'auc'],
                 custom_sep_stats=None,
                 custom_test_stats=None,
                 alpha=0.05, n_jobs=None):

        self.B = int(B)
        self.clf = clf
        self.n_jobs = n_jobs
        if type(separation_stats) == str:
            separation_stats = [separation_stats]
        self.separation_stats = separation_stats
        self.alpha = float(alpha)
        self.custom_sep_stats = custom_sep_stats
        self.custom_test_stats = custom_test_stats

    def get_params(self):
        return {'B': self.B, 'clf': self.clf,
                'custom_sep_stats': self.custom_sep_stats,
                'custom_test_stats': self.custom_test_stats,
                'alpha': self.alpha, 'n_jobs': self.n_jobs}

    def __repr__(self):
        r = 'Two class DiProPerm'
        if hasattr(self, 'metadata'):
            cats = self.classes_
            r += ' of {} vs. {} \n'.format(cats[0], cats[1])
            for s in self.stat:
                r += '{}: {}\n'.format(s, self.test_stats_[s])

        return r

    def save(self, fpath, compress=9):
        dump(self, fpath, compress=compress)

    @classmethod
    def load(cls, fpath):
        return load(fpath)

    def compute_scores(self, X, y):
        """
        Returns the scores of a linear classifier fit on X, y.

        Parameters
        ----------
        X: array-like, shape (n_samples, n_features)
            The X data matrix.

        y: array-list, shape (n_samples,)
            The binary class labels.
        """
        if self.clf == 'md':
            w = get_md_normal_vector(X, y)
        elif callable(self.clf):
            w = self.clf(X, y)
        else:
            raise ValueError("{} is invalid method. Expected: 'md' or callable")

        w = w.reshape(-1)
        w /= np.linalg.norm(w)
        return np.dot(X, w)

    def fit(self, X, y):
        """
        Fits DiProPerm.

        Parameters
        ----------
        X: array-like, shape (n_samples, n_features)
            The X training data matrix.

        y: array-like, shape (n_samples, )
            The observed class labels. Must be binary classes.

        """

        # check arguments
        # X, y = check_X_y(X, y)
        X = np.array(X)
        if X.ndim == 1:
            X = X.reshape(-1, 1)
        y = np.array(y)
        self.classes_ = np.unique(y)
        assert len(self.classes_) == 2
        self.obs_y_ = y
        self.metadata_ = {'counter':  dict(Counter(y)),
                          'shape': X.shape}

        self.obs_scores_ = self.compute_scores(X, y)
        self.perm_scores_, self.perm_y_ = self._compute_perm_scores(X, y)

        self.obs_sep_stats_ = {}
        self.perm_sep_stats_ = {}
        self.test_stats_ = {}
        self.compute_statistics()
        return self

    def compute_statistics(self):
        """

        Computes separation and test statistics after permutation and observed
        scores have been computed.
        """
        assert hasattr(self, 'obs_scores_') and hasattr(self, 'perm_scores_')

        for stat in self.separation_stats:
            obs_sep = get_separation_statistic(self.obs_scores_, self.obs_y_,
                                               stat=stat)

            perm_sep = np.array([get_separation_statistic(self.perm_scores_[b, :],
                                                          self.obs_y_, stat=stat)
                                 for b in range(self.B)])

            tst = get_test_statistics(obs_sep, perm_sep, alpha=self.alpha,
                                      custom_test_stats=self.custom_test_stats)

            self.obs_sep_stats_[stat] = obs_sep
            self.perm_sep_stats_[stat] = perm_sep
            self.test_stats_[stat] = tst

        # user provided separation statistics
        if self.custom_sep_stats is not None:
            for stat, f in self.custom_sep_stats:
                obs_sep = f(scores=self.obs_scores_, y=self.obs_y_)
                perm_sep = np.array([f(scores=self.perm_scores_[b, :], y=self.obs_y_)
                                     for b in range(self.B)])

                tst = get_test_statistics(obs_sep, perm_sep, alpha=self.alpha,
                                          custom_test_stats=self.custom_test_stats)

                self.obs_sep_stats_[stat] = obs_sep
                self.perm_sep_stats_[stat] = perm_sep
                self.test_stats_[stat] = tst

    def _sample_perm(self, X, y):
        """
        Samples scores from one permutation.

        Output
        ------
        perm_scores, perm_y

        perm_scores: array-like, shape (n_samples, )
            The scores computed from the permutation distribution.

        perm_y: array-like, shape (n_samples, )
            The permuted class labels.

        """
        perm_y = np.random.permutation(y)
        perm_scores = self.compute_scores(X, perm_y)
        return perm_scores, perm_y

    def _compute_perm_scores(self, X, y):
        """
        Samples permutation scores.
        """

        # compute permutation statistics in parallel
        if self.n_jobs is not None:
            perm = Parallel(n_jobs=self.n_jobs)\
                    (delayed(self._sample_perm)(X, y) for i in range(self.B))

        else:
            perm = [self._sample_perm(X, y) for _ in range(self.B)]

        perm_scores, y_perm = tuple(zip(*perm))
        return np.array(perm_scores), np.array(y_perm)

    def plot_perm_sep_stats(self, stat):
        """
        Plots a histogram of the DiProPerm null distribution with the
        observed value and some test statistics.

        Parameters
        ----------
        stat: str
            Which separation statistic to show.
        """
        assert stat in self.obs_sep_stats_.keys()

        plot_perm_sep_stats(obs=self.obs_sep_stats_[stat],
                            perm=self.perm_sep_stats_[stat],
                            stat=stat,
                            rejected=self.test_stats_[stat]['rejected'],
                            p=self.test_stats_[stat]['pval'],
                            Z=self.test_stats_[stat]['Z'],
                            cuttoff=self.test_stats_[stat]['cutoff_val'],
                            B=self.B,
                            alpha=self.alpha)

    def plot_observed_scores(self):
        """
        Plots a histogram of the observed scores for the training data.
        """
        plot_observed_scores(self.obs_scores_, self.obs_y_)

    def plot_perm_scores(self, b=0):
        """
        Plots a histogram of the permuted scores for one of the permutation
        samples.

        Parameters
        ----------
        b: int
            Index of the permutation sample to plot.
        """
        plot_perm_scores(perm_scores=self.perm_scores_[b, :],
                         perm_y=self.perm_y_[b, :], obs_y=self.obs_y_)

    def plot_diagnostic(self, stat):
        """
        Plots DiProPerm diagnostic for a given separation statistics.

        - (top left) Observed data scores, colored by observed classes.
        - (top right) DiProPerm null distribution.
        - (bottom left) Permuted data scores of the permutation with the
        smallest statistic value.
        - (bottom right) Permuted data scores of the permutation with the
        largest statistic value.

        Parameters
        ----------
        stat: str
            Which separation statistic.
        """
        # plt.figure(figsize=[20, 20])
        plt.subplot(2, 2, 1)
        self.plot_observed_scores()

        plt.subplot(2, 2, 2)
        self.plot_perm_sep_stats(stat)

        plt.subplot(2, 2, 3)
        idx_min = np.argmin(self.perm_sep_stats_[stat])
        val_min = self.perm_sep_stats_[stat][idx_min]
        self.plot_perm_scores(idx_min)
        plt.title('smallest {} ({:1.3f})'.format(stat, val_min))

        plt.subplot(2, 2, 4)
        idx_max = np.argmax(self.perm_sep_stats_[stat])
        val_max = self.perm_sep_stats_[stat][idx_max]
        self.plot_perm_scores(idx_max)
        plt.title('largest {} ({:1.3f})'.format(stat, val_max))


def get_test_statistics(obs_stat, perm_samples, alpha=0.05,
                        custom_test_stats=None):
    """
    obs_stat: float
        The observed statistic.

    perm_samples: list
        The sampled permutation statistics.

    alpha: float, between 0 and 1
        The cutoff value.

    custom_tests_stats: None, iterable
        User provided custom test statistics. Must allow for
        for stat, f in custom_test_stats where stat is a str
        and f is a function f(obs_stat, perm_samples) returning a value.
    """

    stats = {}
    stats['obs'] = obs_stat

    stats['pval'] = np.mean(obs_stat < perm_samples)
    stats['rejected'] = stats['pval'] < alpha
    stats['cutoff_val'] = np.percentile(perm_samples, 100 * (1 - alpha))

    stats['Z'] = (obs_stat - np.mean(perm_samples)) / np.std(perm_samples)

    if custom_test_stats is not None:
        for stat, f in custom_test_stats:
            stats[stat] = f(obs_stat=obs_stat, perm_samples=perm_samples)

    return stats


def get_separation_statistic(scores, y, stat='md'):
    y = np.array(y)
    classes = np.unique(y)
    assert len(classes) == 2
    s0 = scores[y == classes[0]]
    s1 = scores[y == classes[1]]

    if stat == 'md':
        return abs(np.mean(s0) - np.mean(s1))
    elif stat == 't':
        return abs(ttest_ind(s0, s1, equal_var=False).statistic)
    elif stat == 'auc':
        return roc_auc_score(y == classes[0], scores)
    else:
        raise ValueError("'{} is invalid statistic. Expected one of 'md' or 't'".format(stat))
